compiler/src/org.graalvm.compiler.loop.phases/src/org/graalvm/compiler/loop/phases/ReassociateInvariantPhase.java

```diff
@@ -31,25 +31,36 @@
 import org.graalvm.compiler.nodes.calc.BinaryArithmeticNode;
 import org.graalvm.compiler.phases.Phase;
 
+import java.util.function.Predicate;
+
 /**
  * Rearrange {@link BinaryArithmeticNode#isAssociative() associative binary operations} so that
  * invariant parts of the expression can move outside of the loop.
  */
 public class ReassociateInvariantPhase extends Phase {
+    private static final Predicate<LoopEx> reassociateVariants = loop -> loop.reassociateVariants();
+    private static final Predicate<LoopEx> reassociateInvariants = loop -> loop.reassociateInvariants();
 
     @SuppressWarnings("try")
     @Override
     protected void run(StructuredGraph graph) {
-        int iterations = 0;
         DebugContext debug = graph.getDebug();
-        try (DebugContext.Scope s = debug.scope("ReassociateInvariants")) {
+        try (DebugContext.Scope s = debug.scope("ReassociateLoopInvariant")) {
+            doReassociate(graph, reassociateVariants, debug, "ReassociateVariants");
+            doReassociate(graph, reassociateInvariants, debug, "ReassociateInvariants");
+        } catch (Throwable e) {
+            throw debug.handle(e);
+        }
+    }
+
+    @SuppressWarnings("try")
+    private static void doReassociate(StructuredGraph graph, Predicate<LoopEx> predicate, DebugContext debug, String scope) {
+        int iterations = 0;
+        try (DebugContext.Scope s = debug.scope(scope)) {
             boolean changed = true;
             while (changed) {
-                changed = false;
                 final LoopsData dataReassociate = new LoopsData(graph);
-                for (LoopEx loop : dataReassociate.loops()) {
-                    changed |= loop.reassociateInvariants();
-                }
+                changed = dataReassociate.loops().stream().anyMatch(predicate);
```

Comment:
I would prefer a regular loop here if possible

Reply:
Yeah, a regular loop better. I will update it later. Thanks for the advice!